reconcile/terraform_cloudflare_users.py

```diff
@@ -0,0 +1,380 @@
+        enable_deletion = self.params.enable_deletion
+
+        cloudflare_roles, settings = self._get_desired_state()
+
+        if not settings.settings:
+            raise RuntimeError("App interface setting not defined")
+
+        secret_reader = create_secret_reader(use_vault=settings.settings[0].vault)
+
+        cf_clients = self._build_cloudflare_terraform_config_client_collection(
+            cloudflare_roles, secret_reader, account_name
+        )
+
+        users = get_cloudflare_users(
+            cloudflare_roles.cloudflare_account_roles,
+            account_name,
+            settings.settings[0].cloudflare_email_domain_allow_list,
+        )
+        specs = build_external_resource_spec_from_cloudflare_users(users)
+
+        cf_clients.add_specs(specs)
+        cf_clients.populate_resources()
+
+        working_dirs = cf_clients.dump(print_to_file=print_to_file)
+
+        if print_to_file:
+            return
+
+        # for storing unique CloudflareAccountV1 since cloudflare_account_role_v1 can contain duplicates due to schema
+        account_names_to_account = {
+            role.account.name: role.account
+            for role in cloudflare_roles.cloudflare_account_roles or []
+            if role.account.name in cf_clients.dump()
+        }
+
+        accounts = [
+            acct.dict(by_alias=True) for _, acct in account_names_to_account.items()
+        ]
+
+        self._run_terraform(
+            QONTRACT_INTEGRATION,
+            QONTRACT_INTEGRATION_VERSION,
+            QONTRACT_TF_PREFIX,
+            dry_run,
+            enable_deletion,
+            thread_pool_size,
+            working_dirs,
+            accounts,
+        )
+
+    def _run_terraform(
+        self,
+        QONTRACT_INTEGRATION: str,
+        QONTRACT_INTEGRATION_VERSION: str,
+        QONTRACT_TF_PREFIX: str,
+        dry_run: bool,
+        enable_deletion: bool,
+        thread_pool_size: int,
+        working_dirs: Mapping[str, str],
+        accounts: Iterable[Mapping[str, Any]],
+    ) -> None:
+        tf = TerraformClient(
+            QONTRACT_INTEGRATION,
+            QONTRACT_INTEGRATION_VERSION,
+            QONTRACT_TF_PREFIX,
+            accounts,
+            working_dirs,
+            thread_pool_size,
+        )
+
+        try:
+            disabled_deletions_detected, err = tf.plan(enable_deletion)
+            if err:
+                raise TerraformPlanFailed(
+                    f"Failed to run terraform plan for integration {QONTRACT_INTEGRATION}"
+                )
+            if disabled_deletions_detected:
+                raise TerraformDeletionDetected(
+                    "Deletions detected but they are disabled"
+                )
+
+            if dry_run:
+                return
+
+            err = tf.apply()
+            if err:
+                TerraformApplyFailed(
+                    f"Failed to run terraform apply for integration {QONTRACT_INTEGRATION}"
+                )
+        finally:
```

Comment:
I saw somewhere in the teamchat that sys.exit was used because we needed some way to propagate success/error back to process running on jenkins. Exceptions or lack of exceptions also covers this.